server/embedded/src/org/labkey/embedded/LabKeyServer.java

```diff
@@ -160,32 +160,38 @@ private List<ContextResource> getDataSourceResources(ContextProperties props) th
                     dataSourceResource.setProperty("password", props.getPassword().get(i));
                     dataSourceResource.setProperty("username", props.getUsername().get(i));
 
-                    String maxTotal = props.getMaxTotal().get(i);
-                    maxTotal = maxTotal != null ? maxTotal.trim() : MAX_TOTAL_CONNECTIONS_DEFAULT;
-                    dataSourceResource.setProperty("maxTotal", maxTotal);
-
-                    String maxIdle = props.getMaxIdle().get(i);
-                    maxIdle = maxIdle != null ? maxIdle.trim() : MAX_IDLE_DEFAULT;
-                    dataSourceResource.setProperty("maxIdle", maxIdle);
-
-                    String maxWait = props.getMaxWaitMillis().get(i);
-                    maxWait = maxWait != null ? maxWait.trim() : MAX_WAIT_MILLIS_DEFAULT;
-                    dataSourceResource.setProperty("maxWaitMillis", maxWait);
-
-                    String allowAccess = props.getAccessToUnderlyingConnectionAllowed().get(i);
-                    allowAccess = allowAccess != null ? allowAccess : ACCESS_TO_CONNECTION_ALLOWED_DEFAULT;
-                    dataSourceResource.setProperty("accessToUnderlyingConnectionAllowed", allowAccess);
-
-                    String validationQuery = props.getValidationQuery().get(i);
-                    validationQuery = validationQuery != null ? validationQuery.trim() : VALIDATION_QUERY_DEFAULT;
-                    dataSourceResource.setProperty("validationQuery", validationQuery);
+                    dataSourceResource.setProperty("maxTotal", getPropValue(props.getMaxTotal(), i, MAX_TOTAL_CONNECTIONS_DEFAULT, "maxTotal"));
+                    dataSourceResource.setProperty("maxIdle", getPropValue(props.getMaxIdle(), i, MAX_IDLE_DEFAULT, "maxIdle"));
+                    dataSourceResource.setProperty("maxWaitMillis", getPropValue(props.getMaxWaitMillis(), i, MAX_WAIT_MILLIS_DEFAULT, "maxWaitMillis"));
+                    dataSourceResource.setProperty("accessToUnderlyingConnectionAllowed", getPropValue(props.getAccessToUnderlyingConnectionAllowed(), i, ACCESS_TO_CONNECTION_ALLOWED_DEFAULT, "accessToUnderlyingConnectionAllowed"));
+                    dataSourceResource.setProperty("validationQuery", getPropValue(props.getValidationQuery(), i, VALIDATION_QUERY_DEFAULT, "validationQuery"));
 
                     dataSourceResources.add(dataSourceResource);
                 }
 
                 return  dataSourceResources;
             }
 
+            private String getPropValue(List<String> propList, int index, String defaultValue, String propName)
+            {
+                if (propList == null)
+                {
+                    logger.debug(String.format("%1$s property was not provided, using default", propName));
+                    return defaultValue;
+                }
+
+                try
+                {
+                    String val = propList.get(index);
```

Comment:
It's better to check the list size instead of catching the exception. Just like we should check for nulls instead of catching NullPointerException. Otherwise looks good.

Reply:
Changed to use Map.

Using a List means that any props provided in secondary resources need to also be provided for primary (and/or previously configured resources). The Map seems to bind correctly, and correctly keys the property values.